src/main/java/io/temporal/internal/statemachines/TimerStateMachine.java

```diff
@@ -0,0 +1,179 @@
+import io.temporal.api.command.v1.Command;
+import io.temporal.api.command.v1.StartTimerCommandAttributes;
+import io.temporal.api.enums.v1.CommandType;
+import io.temporal.api.enums.v1.EventType;
+import io.temporal.api.history.v1.HistoryEvent;
+import io.temporal.api.history.v1.TimerCanceledEventAttributes;
+import io.temporal.workflow.Functions;
+
+final class TimerStateMachine
+    extends EntityStateMachineInitialCommand<
+        TimerStateMachine.State, TimerStateMachine.ExplicitEvent, TimerStateMachine> {
+
+  private final StartTimerCommandAttributes startAttributes;
+
+  private final Functions.Proc1<HistoryEvent> completionCallback;
+
+  /**
+   * Creates a new timer state machine
+   *
+   * @param attributes timer command attributes
+   * @param completionCallback invoked when timer fires or reports cancellation. One of
+   *     TimerFiredEvent, TimerCanceledEvent.
+   * @return cancellation callback that should be invoked to initiate timer cancellation
+   */
+  public static TimerStateMachine newInstance(
+      StartTimerCommandAttributes attributes,
+      Functions.Proc1<HistoryEvent> completionCallback,
+      Functions.Proc1<CancellableCommand> commandSink,
+      Functions.Proc1<StateMachine> stateMachineSink) {
+    return new TimerStateMachine(attributes, completionCallback, commandSink, stateMachineSink);
+  }
+
+  private TimerStateMachine(
+      StartTimerCommandAttributes attributes,
+      Functions.Proc1<HistoryEvent> completionCallback,
+      Functions.Proc1<CancellableCommand> commandSink,
+      Functions.Proc1<StateMachine> stateMachineSink) {
+    super(STATE_MACHINE_DEFINITION, commandSink, stateMachineSink);
+    this.startAttributes = attributes;
+    this.completionCallback = completionCallback;
+    explicitEvent(ExplicitEvent.SCHEDULE);
+  }
+
+  enum ExplicitEvent {
+    SCHEDULE,
+    CANCEL
+  }
+
+  enum State {
+    CREATED,
+    START_COMMAND_CREATED,
+    START_COMMAND_RECORDED,
+    CANCEL_TIMER_COMMAND_CREATED,
+    FIRED,
+    CANCELED,
+  }
+
+  public static final StateMachineDefinition<State, ExplicitEvent, TimerStateMachine>
+      STATE_MACHINE_DEFINITION =
+          StateMachineDefinition.<State, ExplicitEvent, TimerStateMachine>newInstance(
+                  "Timer", State.CREATED, State.FIRED, State.CANCELED)
+              .add(
+                  State.CREATED,
+                  ExplicitEvent.SCHEDULE,
+                  State.START_COMMAND_CREATED,
+                  TimerStateMachine::createStartTimerCommand)
+              .add(
+                  State.START_COMMAND_CREATED,
+                  CommandType.COMMAND_TYPE_START_TIMER,
+                  State.START_COMMAND_CREATED)
+              .add(
+                  State.START_COMMAND_CREATED,
+                  EventType.EVENT_TYPE_TIMER_STARTED,
+                  State.START_COMMAND_RECORDED,
+                  EntityStateMachineInitialCommand::setInitialCommandEventId)
+              .add(
```

Comment:
there is zero chance we somehow get a timer fired event before the timer started event? This should be properly serialized on the server, right?

Reply:
No, as timerStarted is written transactionally with the task addition to the timer queue. And due to buffered events the guarantee is even stronger. Command events always follow WorkflowTaskCompleted event.